### PR TITLE
APEXMALHAR-2105 enhancing CSV formatter to read field info from schema

### PR DESCRIPTION
**Pull request by shubham-pathak22**

### CODE REVIEW

**Comment by shubham-pathak22** [comment]
@devtagare @chinmaykolhatkar @sandeepdeshmukh  Please review.

File: contrib/src/main/java/com/datatorrent/contrib/formatter/CsvFormatter.java [hunk 1]
```diff
@@ -58,215 +68,156 @@
...
+   * metric to keep count of number of tuples emitted on error port port
+   */
+  @AutoMetric
+  protected long errorTupleCount;
+
+  /**
+   * metric to keep count of number of tuples emitted on out port
+   */
+  @AutoMetric
+  protected long emittedObjectCount;
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/formatter/CsvFormatter.java]: Can you please make this private?

File: contrib/src/main/java/com/datatorrent/contrib/formatter/CsvFormatter.java [hunk 2]
```diff
@@ -58,215 +68,156 @@
...
 
-  protected transient String[] nameMapping;
-  protected transient CellProcessor[] processors;
-  protected transient CsvPreference preference;
+  /**
+   * metric to keep count of number of tuples emitted on input port
+   */
+  @AutoMetric
+  protected long incomingTuplesCount;
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/formatter/CsvFormatter.java]: Can you make this private?

File: contrib/src/main/java/com/datatorrent/contrib/formatter/CsvFormatter.java [hunk 3]
```diff
@@ -58,215 +68,156 @@
...
 
   public CsvFormatter()
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/formatter/CsvFormatter.java]: You can remove the constructor as nothing is there in it.

File: contrib/src/main/java/com/datatorrent/contrib/formatter/CsvFormatter.java [hunk 4]
```diff
@@ -58,215 +68,156 @@
...
   }
 
-  /**
-   * Sets the delimiter which separates fields in incoming data.
-   * 
-   * @param fieldDelimiter
-   */
-  public void setFieldDelimiter(int fieldDelimiter)
+  @VisibleForTesting
+  public long getEmittedObjectCount()
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/formatter/CsvFormatter.java]: You can make this protected? As the testcase will be in the same package, it'll be accesible via protected as well.

File: contrib/src/main/java/com/datatorrent/contrib/formatter/CsvFormatter.java [hunk 5]
```diff
@@ -58,215 +68,156 @@
...
-  /**
-   * Sets the name of the fields with type and format in data as comma separated
-   * string in same order as incoming data. e.g
-   * name:string,dept:string,eid:integer,dateOfJoining:date|dd/mm/yyyy
-   * 
-   * @param fieldInfo
-   */
-  public void setFieldInfo(String fieldInfo)
+  @VisibleForTesting
+  public long getIncomingTuplesCount()
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/formatter/CsvFormatter.java]: Please make this protected... No need for this to be private.

**Comment by shubham-pathak22** [comment]
incorporated review comments.

**Comment by chinmaykolhatkar** [comment]
@shubham-pathak22 CsvFormatterTest is failing for me. Can you please check at your end?

**Comment by chinmaykolhatkar** [comment]
Also can you please add ApplicationTest for this operator? That will ensure that the serialization related to the operator is happening correctly.

**Comment by chinmaykolhatkar** [comment]
@shubham-pathak22 Can you please rebase this PR with latest master, add ApplicationTest and run CsvFormatterTest manually from cmd line?

There is an important fix that went to contrib.

**Comment by shubham-pathak22** [comment]
Incorporated review comments

**Comment by shubham-pathak22** [comment]
This has been merged by @chinmaykolhatkar . Closing this PR

**Pull request closed without merging**